go/oasis-node/cmd/debug/txsource/workload/queries.go

```diff
@@ -0,0 +1,455 @@
+	// is special cased with increased probability.
+	var round uint64
+	p := rng.Float32()
+	switch {
+	case p < queriesEarliestHeightRatio:
+		round = 1
+	default:
+		// [1, latestRound]
+		round = uint64(rng.Int63n(int64(latestRound) + 1))
+	}
+
+	// GetBlock.
+	block, err := q.runtime.GetBlock(ctx, &runtimeClient.GetBlockRequest{
+		RuntimeID: q.runtimeID,
+		Round:     round,
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.GetBlock, round: %d: %w", round, err)
+	}
+	block2, err := q.runtime.GetBlockByHash(ctx, &runtimeClient.GetBlockByHashRequest{
+		RuntimeID: q.runtimeID,
+		BlockHash: block.Header.EncodedHash(),
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.GetBlockByHash, hash: %s: %w", block.Header.EncodedHash(), err)
+	}
+	if block.Header.EncodedHash() != block2.Header.EncodedHash() {
+		return fmt.Errorf("Expected equal blocks, got: byRound: %s byHash: %s", block.Header.EncodedHash(), block2.Header.EncodedHash())
+	}
+
+	_, err = q.runtime.QueryTxs(ctx, &runtimeClient.QueryTxsRequest{
+		RuntimeID: q.runtimeID,
+		Query: runtimeClient.Query{
+			RoundMin: 0,
+			RoundMax: round,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.QueryTxs: %w", err)
+	}
+
+	q.logger.Debug("Done runtime queries",
+		"latest_round", latestRound,
+		"block", block,
+		"round", round,
+	)
+
+	return nil
+}
+
+func (q *queries) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, fundingAccount signature.Signer) error {
+	ctx := context.Background()
+
+	q.logger = logging.GetLogger("cmd/txsource/workload/queries")
+
+	q.consensus = cnsc
+	q.registry = registry.NewRegistryClient(conn)
+	q.runtime = runtimeClient.NewRuntimeClient(conn)
+	q.scheduler = scheduler.NewSchedulerClient(conn)
+	q.staking = staking.NewStakingClient(conn)
+
+	// TODO: could add a methods to get consensus parameters directly.
+	doc, err := q.consensus.StateToGenesis(ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("consensus.StateToGenesis error: %w", err)
+	}
+	q.epochtimeParams = doc.EpochTime.Parameters
+	q.stakingParams = doc.Staking.Parameters
+	q.schedulerParams = doc.Scheduler.Parameters
+
+	// Setup simple-keyvalue runtime info.
+	err = q.runtimeID.UnmarshalHex(viper.GetString(CfgRuntimeID))
+	if err != nil {
+		q.logger.Error("runtime unmarshal error",
+			"err", err,
+			"runtime_id", viper.GetString(CfgRuntimeID),
+		)
+		return fmt.Errorf("Runtime unmarshal: %w", err)
+	}
+
+	for {
+		block, err := q.consensus.GetBlock(ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("consensus.GetBlock error: %w", err)
+		}
+
+		// Select height at which queries should be done. Earliest (height=1)
+		// is special cased with increased probability.
+		var height int64
+		p := rng.Float32()
+		switch {
+		case p < queriesEarliestHeightRatio:
+			height = 1
+		default:
+			// [1, block.Height]
+			height = rng.Int63n(block.Height) + 1
```

Comment:
we don't exercise passing `0` for the latest. probably due to the routines doing multiple queries with this parameter and requiring some consistency

Reply:
Yeah and also since that should be (implicitly at least) already tested in other workloads / sanity checks. see: https://github.com/oasislabs/oasis-core/pull/2769#discussion_r395551939

Reply:
resolving with the above thread